lib/Dialect/ESI/runtime/cpp/include/esi/Ports.h

```diff
@@ -51,21 +57,77 @@ class WriteChannelPort : public ChannelPort {
   virtual void write(const MessageData &) = 0;
 };
 
-/// A ChannelPort which reads data from the accelerator.
+/// A ChannelPort which reads data from the accelerator. It has two modes:
+/// Callback and Polling which cannot be used at the same time. The mode is set
+/// at connect() time. To change the mode, disconnect() and then connect()
+/// again.
 class ReadChannelPort : public ChannelPort {
+
 public:
-  using ChannelPort::ChannelPort;
+  ReadChannelPort(const Type *type) : ChannelPort(type) {}
+  virtual void disconnect() override { mode = Mode::Disconnected; }
```

Comment:
With these split modes, and different modes having different member variables associated with them, i'd personally split those into two separate `ReadChannelPort`-derived classes to avoid any confusion/misuse/leak between the two API modes.

Reply:
This is something I considered. The problem is that these channel ports are constructed when the system gets built from the manifest, not when the user locates them. So the user doesn't have the chance to specify the type before getting the reference to them.